events/handler.go

```diff
@@ -0,0 +1,177 @@
+
+const (
+	OomInstanceCat EventCategory = iota
+	ContainerCreation
+	ContainerDeletion
+)
+
+// a general interface which populates the Event field EventData. The actual
+// object, such as an OomInstance, is set as an Event's EventData
+type EventDataInterface interface {
+}
+
+// returns a pointer to an initialized Events object
+func NewEventManager() *events {
+	return &events{
+		eventlist: make(EventSlice, 0),
+	}
+}
+
+// returns a pointer to an initialized Request object
+func NewRequest() *Request {
+	return &Request{
+		EventType: map[EventCategory]bool{},
+	}
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Len() int {
+	return len(e)
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Swap(i, j int) {
+	e[i], e[j] = e[j], e[i]
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Less(i, j int) bool {
+	return e[i].TimeOfEvent.Before(e[j].TimeOfEvent)
+}
+
+// sorts and returns up to the last N chronological elements
+func getLastNEvents(request *Request, eSlice EventSlice) EventSlice {
+	sort.Sort(eSlice)
+	n := request.LastNEvents
+	if n >= eSlice.Len() || n <= 0 {
+		return eSlice
+	}
+	return eSlice[eSlice.Len()-n:]
+}
+
+// determines if an event occurs within the time set in the request object and is the right type
+func checkIfEventSatisfiesRequest(request *Request, event *Event) bool {
+	startTime := request.StartTime
+	endTime := request.EndTime
+	eventTime := event.TimeOfEvent
+	if !startTime.IsZero() {
+		if startTime.After(eventTime) {
+			return false
+		}
+	}
+	if !endTime.IsZero() {
+		if endTime.Before(eventTime) {
+			return false
+		}
+	}
+	if request.EventType[event.EventType] != true {
+		return false
+	}
+	return true
+}
+
+// method of Events object that screens Event objects found in the eventlist
+// attribute and if they fit the parameters passed by the Request object,
+// adds it to a slice of *Event objects that is returned
+func (self *events) GetEvents(request *Request) (EventSlice, error) {
+	returnEventList := EventSlice{}
+	for _, e := range self.eventlist {
+		if checkIfEventSatisfiesRequest(request, e) {
+			returnEventList = append(returnEventList, e)
+		}
+	}
+	returnEventList = getLastNEvents(request, returnEventList)
+	return returnEventList, nil
+}
+
+// method of Events object that maintains an *Event channel passed by the user.
+// When an event occurs that satisfies the parameters in the passed Request
+// object it is fed to the channel
+func (self *events) WatchEvents(inChannel chan *Event, outChannel chan *Event, request *Request) error {
+	for event := range inChannel {
+		if checkIfEventSatisfiesRequest(request, event) {
+			outChannel <- event
+		}
+	}
+	return nil
+}
+
+// method of Events object that adds the argument Event object to the
+// Events.eventlist
+func (self *events) AddEvent(e *Event, passToWatch chan *Event) error {
```

Comment:
We can abstract away the passToWatch and remove it from WatchEvents() by storing them locally. When someone registers a watch we add it to a slice of chan *Event. We can then look through those when we add an event. Would make it simpler for the caller so they don't have to remember the channel